# Task 1

Write a Python function called sqrt2 that calculates and
prints to the screen the square root of 2 to 100 decimal places. Your code should
not depend on any module from the standard library1 or otherwise. You should
research the task first and include references and a description of your algorithm.

# Code

In [4]:
def sqroot():

    # set a value of x
    x = 2
    #storing a value for 10**100 to be used in the while loop
    power = 10**100
    # getting a value for 2 that is 100 places above the decimal point
    x *=(10**(200))
    #initialize y
    y = 0
    
    # A loop to generate the square root of 2 value
    while y!=power:
        y = power
        power = (y + ( x // y)) >> 1
        
    # variable value stores a list of the while loop output
    value = list(str(power))
    
    # insert a dot after 1 place in the list
    value.insert(1, '.')
    
    # store the list values in a variable to be returned to user of the function
    ans = ''.join(value)
    
    # return the answer
    return ans

# call function
sqroot()

10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000


'1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727'

# Analysis

For this task I created a function called sqroot to output the result.
The x value is 2.The x variable is then times by 10 beforeit is exponentiated by 200 to get a value for x that is 100 decimals places above the decimal point.A variable y is created for the while loop and initialized at 0. A variable power is set to 10  exponentiated by 100. The while loop runs until y = power. Inside the while loop y is set to power while power is set to y + (x divided y) before being converted into a binary format and then the operations will be performed bit by bit.
The result of the while loop is soted in power before being added cast to a string and added to a list called value. A dot seperator is added to be used after the first value in the list.
The list value variables are then saved in a variable called ans. The ans variable is returned at the end of the function. 

# References

Powers function in python:https://docs.python.org/3/tutorial/introduction.html <br />
Square root:https://www.tutorialspoint.com/How-to-perform-square-root-without-using-math-module-in-Python <br />
formating python:https://www.geeksforgeeks.org/precision-handling-python/ <br />
code: https://stackoverflow.com/questions/64278117/is-there-a-way-to-create-more-decimal-points-on-python-withoutimporting-a-libra <br />
Undestanding Newtons method: https://www.math.ubc.ca/~pwalls/math-python/roots-optimization/newton/ <br />
    https://www.youtube.com/watch?v=2158QbsunA8&ab_channel=LorenzoSadun <br />
To understand how to add a dot seperator to list: https://realpython.com/python-string-split-concatenate-join/ <br />
Generating a square root value: https://stackoverflow.com/questions/5187664/generating-digits-of-square-root-of-2

# Task 2

The Chi-squared test for independence is a statistical hypothesis test like a t-test. It is used to analyse whether two categorical variables are independent. The Wikipedia article gives the table below as an example [4], stating the Chi squared value based on it is approximately 24.6. Use scipy.stats to verify this value and calculate the associated p value. You should include a short note with references justifying your analysis in a markdown cell.

# Table

In [9]:
headings = ["A","B","C","D"]
data =[[ 90, 60, 104, 95],[ 30, 50, 51, 20],[30, 40, 45, 35]]
format_row = "{:>10}" * (len(headings) + 1)
print(format_row.format("", *headings))
for team, row in zip(headings, data):
    print(format_row.format(team, *row)) 

                   A         B         C         D
         A        90        60       104        95
         B        30        50        51        20
         C        30        40        45        35


# Code

In [ ]:
import scipy.stats as stats
  
# data stored in a 2-D array from the table provided
data = [[ 90, 60, 104, 95],[ 30, 50, 51, 20],[30, 40, 45, 35]]
#set a significant value
sigValue = 24.6
stat, p, dof, expected = stats.chi2_contingency(data) 

print("p value is " + str(p)) 

print("Expected")
print(expected)

print("Dof = ", dof)
print("Stat = ", stat)

# Analysis

The result received in this test is 24.57. This is lower than the significant value of 24.6. The difference is 0.03. A null hypothesis can now not be accepted because we can see that there is a relationship between the 2 categories. The value we get for p also reaffirms this conclusion.

# References

Undestanding Chi-Square test: https://stattrek.com/chi-square-test/independence.aspx
https://www.investopedia.com/terms/c/chi-square-statistic.asp#:~:text=A%20chi%2Dsquare%20(%CF%872,compares%20to%20actual%20observed%20data.&text=The%20chi%2Dsquare%20statistic%20compares,of%20variables%20in%20the%20relationship)
Understanding desgrees of freedom:https://statisticsbyjim.com/hypothesis-testing/degrees-freedom-statistics/
Scipy docs:https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chisquare.html

# Task 3

The standard deviation of an array of numbers x is calculated using numpy as np.sqrt(np.sum((x - np.mean(x))** 2)/len(x)).
However, Microsoft Excel has two different versions of the standard deviation calculation, STDEV.P and STDEV.S .
The STDEV.P function performs the above calculation but in the STDEV.S calculation the division is by len(x)-1 rather
than len(x) . Research these Excel functions, writing a note in a Markdown cell about the difference between them. Then use numpy to perform a simulation demonstrating that the STDEV.S calculation is a better estimate for the standard deviation of a population when performed on a sample. 
Note that part of this taskis to figure out the terminology in the previous sentence.

# Research

Understanding Stdev in excel:https://www.youtube.com/watch?v=NdIVwuwfMIg&ab_channel=DesigneLearningTutorials <br />
http://mathbitsnotebook.com/Algebra1/StatisticsData/STPopSample.html <br />
Stdev.p is used to find the standard deviation of all the data acquired. The p stands for population. In excel, you run the formula while selecting all data cells in the spreadsheet to find the result.
Stdev.s is used to calculate the standard deviation of a sample of the data required. In excel you choose a select number of data cells and run the formula to find the standard deviation of those selected cells.

# Code

# Standard Deviation Population Test

In [21]:
import numpy as np
#data is random results of a test in % format - 25 in total
data = ([75,22,82,48,51,92,72,64,38,57,62,79,88,28,53,83,75,66,43,52,78,91,85,34,58])
print("standard deviation population result")
#formula for sample - np.sqrt(np.sum((x - np.mean(x))**2)/len(x))
np.sqrt(np.sum((data - np.mean(data))**2)/len(data))

standard deviation population result


19.777724843874232

# Standard Deviation Sample Test

In [5]:
import numpy as np
#data is random results of a test in % format - 25 in total
data = ([75,22,82,48,51,92,72,64,38,57,62,79,88,28,53,83,75,66,43,52,78,91,85,34,58])
print("standard deviation sample result")
np.sqrt(np.sum((data - np.mean(data))**2)/len(data)-1)

standard deviation sample result


19.75242769889312

# Analysis

The results of both tests gives the user a basis of where the standard deviation of the tests is based. Stdev.p = 19.78(to 2 decimal places) while stdev.s = 17.75(to two decimal places). The difference to two decimal places is 0.03. This gives the user an understanding that the data is mainly between 20% of the mean.
Using Stdev.s gives the user a better understanding of what the standard deviaton is because it will offer a greater fluctuation of results due it taking only select results that are randomly choosen from the data provided.When the data provided gets larger, the population result can change rapidly because alot of people could score around the same mark.This will skew the data and give out a result that may not accurately reflect the data provided to give the best outcome. Due to this using the sample standard deviation formula allows the data to be randomly choosen and give a better outcome because the data provided will be of a greater variance. 
With the data provided in the above code size was not 25 but 2500 and 5-% of the results are between 45-55, this will cause the standard dev population to be a much lower number than the number calculated in this file.

# Task 4

NB – when I first posted this task, I accidentally
wrote “k-means” where I meant to write “kNN” for k Nearest Neighbours. Because of this, I will allow either algorithm to be used and have
extended the deadline by two weeks. Use scikit-learn to apply k-means
clustering to Fisher’s famous Iris data set. You will easily obtain a copy of the data
set online. Explain in a Markdown cell how your code works and how accurate it
might be, and then explain how your model could be used to make predictions of
species of iris

# Code

In [3]:
from sklearn.cluster import KMeans
import numpy as np
#Fisher's Iris Dataset
iris = datasets.load_iris()

X = np.array([[1, 2], [1, 4], [1, 0],
              [10, 2], [10, 4], [10, 0]])
kmeans = KMeans(n_clusters=2, random_state=0).fit(X)
kmeans.labels_

kmeans.predict([[0, 0], [12, 3]])

kmeans.cluster_centers_

array([[10.,  2.],
       [ 1.,  2.]])